### PR TITLE
add constants and binary utils for binary trie

### PR DESCRIPTION
**Pull request by NIC619**
### How was it fixed?
Add binary trie part 1: Add the constants and the binary encoding/decoding utilities(and tests) for binary trie

#### Cute Animal Picture

Cute animal picture

### CODE REVIEW

File: trie/utils/binaries.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+from trie.constants import (
+    TWO_BITS,
+    PREFIX_00,
+    PREFIX_100000,
+)
+
+
+def safe_ord(value):
+    if isinstance(value, int):
```
Comment (pipermerriam) [inline trie/utils/binaries.py]: FYI, `isinstance(value, int)` will return truthy for `True` and `False`.  You can use `eth_utils.is_integer` if you need that kind of precision.

**Comment by NIC619** [comment]
@pipermerriam Thanks for the reminder. Replaced it with `is_integer` as a precaution.

File: trie/utils/binaries.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        return value
+    else:
+        return ord(value)
+
+
+def decode_from_bin(input_bin):
```
Comment (pipermerriam) [inline trie/utils/binaries.py]: I just toyed with an alternate implementation of this.

```python
from eth_utils import (
    apply_to_return_value,
)
from cytoolz import (
    partition_all,
)


@apply_to_return_value(bytes)
def alt_decode_from_bin(input_bin):
    for chunk in partition_all(8, input_bin):
        yield sum(
            2**exp * bit
            for exp, bit
            in enumerate(reversed(chunk))
        )
```

Some *very* rudimentary benchmarking suggests it's about 30% faster than the current implementation.

File: trie/utils/binaries.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    output = bytearray(len(input_bin) // 8)
+    for i in range(0, len(input_bin), 8):
+        byte_value = 0
+        for bit_value in input_bin[i:i+8]:
+            byte_value = byte_value * 2 + bit_value
+        output[i//8] = byte_value
+    return bytes(output)
+
+
+def encode_to_bin(value):
```
Comment (pipermerriam) [inline trie/utils/binaries.py]: Ended up also taking a stab at an alternate implementation of this.

```python
EXP = tuple(reversed(tuple(2**i for i in range(8))))


@apply_to_return_value(bytes)
def alt_encode_to_bin(value):
    for char in value:
        for exp in EXP:
            if char & exp:
                yield True
            else:
                yield False
```

Benchmarking suggests it's 5-20% faster than the current implementation and I think it's a bit easier to read as well.

**Review by pipermerriam — CHANGES_REQUESTED** [review]
I got curious and ended up trying out some ideas that I think improve both readability *and* performance.  Lemme know what you think.

**Review by pipermerriam — APPROVED** [review]

**Comment by NIC619** [comment]
Though I didn't go too much into performance part, it definitely looks much cleaner. Will definitely try to adhere to this pattern.

**Comment by NIC619** [comment]
Merged with new binary encoding/decoding.

**Pull request merged**